hera_cal/redcal.py

```diff
@@ -618,8 +612,61 @@ def firstcal(self, data, df, wgts={}, sparse=False, mode='default', norm=True, m
             w_ls[eq_key] = twgts[(bl1, bl2)]
         ls = linsolve.LinearSolver(d_ls, wgts=w_ls, sparse=sparse)
         sol = ls.solve(mode=mode)
-        sol = {self.unpack_sol_key(k): v[0] for k, v in sol.items()}  # ignoring offset
-        return sol
+        dly_sol = {self.unpack_sol_key(k): v[0] for k, v in sol.items()}
+        off_sol = {self.unpack_sol_key(k): v[1] for k, v in sol.items()}
+        return dly_sol, off_sol
+
+    def firstcal(self, data, freqs, wgts={}, maxiter=25, conv_crit=1e-6,
+                 sparse=False, mode='default', norm=True, medfilt=False, kernel=(1, 11)):
+        """Solve for a calibration solution parameterized by a single delay and phase offset
+        per antenna using the phase difference between nominally redudant measurements. 
+        Delays are solved in a single iteration, but phase offsets are solved for 
+        iteratively to account for phase wraps.
+
+        Args:
+            data: visibility data in the dictionary format {(ant1,ant2,pol): np.array}
+            freqs: numpy array of frequencies in the data
+            wgts: dictionary of linear weights in the same format as data. Defaults to equal wgts.
+            maxiter: maximum number of phase offset solver iterations
+            conv_crit: convergence criterion for iterative offset solver, defined as the L2 norm
+                of the changes in phase (in radians) over all times and antennas
+            sparse: represent the A matrix (visibilities to parameters) sparsely in linsolve
+            mode: solving mode passed to the linsolve linear solver ('default', 'lsqr', 'pinv', or 'solve')
+                Suggest using 'default' unless solver is having stability (convergence) problems.
+                More documentation of modes in linsolve.LinearSolver.solve().
+            norm: calculate delays from just the phase information (not the amplitude) of the data.
+                This is a pretty effective way to get reliable delay even in the presence of RFI.
+            medfilt : boolean, median filter data before fft.  This can work for data containing
+                unflagged RFI, but tends to be less effective in practice than 'norm'.  Default False.
+            kernel : size of median filter kernel along (time, freq) axes
+
+        Returns:
+            g_fc: dictionary of Ntimes x Nfreqs per-antenna gains solutions in the 
+                {(index, antpol): np.exp(2j * np.pi * delay * freqs + 1j * offset)} format.
+        """
+        # Run firstcal with both delay and offset solving:
+        df = np.median(np.ediff1d(freqs))
+        dtype = np.find_common_type([d.dtype for d in data.values()], [])
+        dly_fc, off_fc = self._firstcal_iteration(data, df=df, f0=freqs[0], wgts=wgts, sparse=sparse, 
```

Comment:
Why have a first iteration outside the for loop?  Can this be avoided?

Reply:
Yes, it can. Done.